### DISCUSSION TITLE
Applying textures on models from .obj files

### DISCUSSION THREAD

**Discussion by cardle7445** [pyvista/pyvista#2622] [Q&A]
## Description
Hello,
I found PyVista an amazing module that helps a lot with my research. Now I'm facing a problem of texture mapping onto my .obj mesh. I produced several .obj files and texture images using contextcapture, and I would like to display them together using PyVista.

```
import pyvista as pv
from pyvistaqt import BackgroundPlotter
pl = BackgroundPlotter()
for i in range(1,10):
    obj = "D:\\Data\\Tile_{}\\Tile_{}.obj".format(i, i)
    tex_image = "D:\\Data\\Tile_{}\\Tile_{}_0.jpg".format(i, i)
    tex = pv.read_texture(tex_image)
    mesh = pv.read(obj)
    pl.add_mesh(mesh, texture=tex)
```

## Example Data

Here is what I got:
![image](https://user-images.githubusercontent.com/82077466/135326706-1f892a15-4339-4706-8de0-0aab10724293.png)

It seems that most of the .obj files successfully mapped with the texture images, but some of them didn't. 

But if I open it with windows 3D viewer, there's no any problem:
![image](https://user-images.githubusercontent.com/82077466/135330051-3efa162f-4ead-403d-9410-cc1629b3c34c.png)

Here's a link to the .obj file and texture image that have the above-mentioned issue, thanks.
https://drive.google.com/file/d/1Kde4RQCNvYOzdNjf0Vy7G3X5Exf6_ukT/view?usp=sharing

**Comment by 109021017** [comment]
There's a texture named 'Tile_4_untextured' declared in your mtl file which is not used by any faces in your obj. Remove that in both .mtl and .obj then you can view it without any problem.

**Comment by cardle7445** [comment]
@109021017 Thanks a lot! I just figured out where the problem's at and have already solved it a few days ago.

Actually, in all of my .mtl files I found there's a texture named 'Tile_?_untextured'  which is not used by any faces. When importing all of the .obj files and applying textures on them, only some of them appear all black. I still don't know the reason why this should happen. But like what you said, if I remove those lines in my .obj files and .mtl file, it works! Thanks again!

**Comment by sharoseali** [comment]
HI Guys. I did point clouds up-sampling which returns me up-sampled mesh or point_cloud with no color information.  Now I have to reproduce their colors. how can i do it. I take some shots of color point cloud and tried to use images as texture
```
tex = pv.read_texture(tex_image)
pl.add_mesh(mesh, texture=tex)
```
but its not working giving me :
```

~/Desktop/point_clouds/Dis-PU$ python texture_mapping_Vista.py 
Traceback (most recent call last):
  File "texture_mapping_Vista.py", line 66, in <module>
    pl.add_mesh(mesh, texture=tex)
  File "/home/casper/anaconda3/lib/python3.7/site-packages/pyvista/plotting/plotting.py", line 1976, in add_mesh
    raise ValueError('Input mesh does not have texture coordinates to support the texture.')
ValueError: Input mesh does not have texture coordinates to support the texture.
```
Anyone Help ? Thanks

**Comment by gurkirt** [comment]
I am trying to load .obj file with one .mtl file and two .jpg images. But above method doesn't wok for me. How to load two texture images for a mesh (.obj) with .mtl file?

**Comment by 109021017** [comment]
@gurkirt I find that `pv.read()` can parse the obj file successfully, with texture names and uv coordinates and everything. The uv is stored in `mesh.point_arrays[material_name]`. But it can plot only one texture (the first one I set) even if I set the `mesh.textures` with different textures. 

So, my solution is split the mesh by the 'MaterialIds' and then add these meshes into one plotter. Here's the code: 

```python
def plot_obj_with_multiple_textures(obj_path, mtl_path=None, texture_dir=None):
    obj_mesh = pv.read(obj_path)
    if mtl_path is None:
        # parse the obj file for mtl_path if the mtl_path is not set.
        pass
    if texture_dir is None:
        texture_dir = os.path.dirname(obj_path)

    texture_paths = []
    mtl_names = []

    # parse the mtl file
    with open(mtl_path) as mtl_file:
        for line in mtl_file.readlines():
            parts = line.strip().split()
            if len(parts) < 2:
                continue
            if parts[0] == 'map_Kd':
                texture_paths.append(os.path.join(texture_dir, parts[1]))
            elif parts[0] == 'newmtl':
                mtl_names.append(parts[1])

    plotter = pv.Plotter()
    material_ids = obj_mesh.cell_arrays['MaterialIds']
    
    # This part is not working.
    # for i in np.unique(material_ids):
    #     obj_mesh.textures[mtl_names[i]] = pv.read_texture(texture_paths[i])
    # plotter.add_mesh(obj_mesh)
    
    # This one do.
    for i in np.unique(material_ids):
        mesh_part = obj_mesh.extract_cells(material_ids == i)
        mesh_part.textures[mtl_names[i]] = pv.read_texture(texture_paths[i])
        plotter.add_mesh(mesh_part)
    plotter.show()
```

**Comment by gurkirt** [comment]
@109021017 it worked like charm, I can not thank you enough!

**Comment by segobytes** [comment]
That solution stopped working in version 0.41: specifically this line:
 mesh_part.textures[mtl_names[i]] = pv.read_texture(texture_paths[i])
now returns an error: AttributeError: 'UnstructuredGrid' object has no attribute 'textures' 
and indeed, in version 0.41, pyvista.DataSet.textures is not documented anymore.
